### PR TITLE
Assignment3

### PR DESCRIPTION
**Pull request by ashwath462**

### CODE REVIEW

File: src/components/Navbar.svelte [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<script lang="ts">
+	import "../app.css";
+	import { goto } from '$app/navigation';
+    import {auth} from '$lib/firebase/firebase'
```
Comment (piyushmalviya13) [inline src/components/Navbar.svelte]: formatting is incorrect

File: src/components/TaskTable.svelte [hunk 2]
```diff
@@ -0,0 +1,93 @@
+<script lang="ts">
+	import "../app.css";
+	import {onMount} from 'svelte';
+	export let type: 'To-Do' | 'Done';
+	import {toDoList, doneList, user} from '$lib/store/store';
+	import { database } from "../utils/utils";
+  	import DeleteButton from "./delete-button.svelte";
+
+	let tasks:any = [], otherTasks:any = [], currentUser:any, currentUserId:any;
```
Comment (piyushmalviya13) [inline src/components/TaskTable.svelte]: why is any type used?
Don't you have data type for task, user and current user id would be a string right.

File: src/components/TaskTable.svelte [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	let obj = {
+		checkboxBg: '',
+		color: ''
+	};
+
+	
+
+	onMount(async()=>{
+		currentUserId = localStorage.getItem('userId');
```
Comment (piyushmalviya13) [inline src/components/TaskTable.svelte]: can we make utility methods for getting and setting values?

File: src/components/TaskTable.svelte [hunk 4]
```diff
@@ -0,0 +1,93 @@
+<script lang="ts">
+	import "../app.css";
+	import {onMount} from 'svelte';
+	export let type: 'To-Do' | 'Done';
+	import {toDoList, doneList, user} from '$lib/store/store';
+	import { database } from "../utils/utils";
+  	import DeleteButton from "./delete-button.svelte";
+
+	let tasks:any = [], otherTasks:any = [], currentUser:any, currentUserId:any;
+	let obj = {
```
Comment (piyushmalviya13) [inline src/components/TaskTable.svelte]: this obj represents what?

File: src/components/TaskTable.svelte [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	onMount(async()=>{
+		currentUserId = localStorage.getItem('userId');
+		if(currentUserId){
+			const userData:any = await database.getUserTasks(currentUserId);
+			console.log({userData});
+			toDoList.set(userData?.toDoList);
+			doneList.set(userData?.doneList);
+			user.set({uid: currentUserId, email: userData?.email});
+		}
+		await user.subscribe((value:any)=>{
```
Comment (piyushmalviya13) [inline src/components/TaskTable.svelte]: why is await used here.
Are you aware how subscribe method works?

File: src/components/TaskTable.svelte (hunk 2 above)
Comment (piyushmalviya13) [inline src/components/TaskTable.svelte]: What is task and what is otherTasks

File: src/components/TaskTable.svelte [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+				otherTasks = value;
+			})
+
+		} else throw new Error('Kindly check type');
+	})
+
+	const deleteTask = (id: number) => {
+		tasks = tasks.filter((task: any) => task.id != id);
+		console.log(tasks,otherTasks);
+		(type === "To-Do")?(
```
Comment (piyushmalviya13) [inline src/components/TaskTable.svelte]: Task types should be an enum

File: src/components/TaskTable.svelte [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+					type="checkbox"
+					checked={task.isComplete}
+					on:click={() => {
+						handleCheck(task.id);
+					}}
+					class="checkbox {obj.checkboxBg} checkbox-xl"
+				/>
+				<div class="w-[35vw]">{task.content}</div>
+				<button on:click={() => deleteTask(task.id)}><DeleteButton/></button>
+			</div>
```
Comment (piyushmalviya13) [inline src/components/TaskTable.svelte]: This should be a separate component TaskTile.svelte

File: src/components/delete-button.svelte [hunk 8]
```diff
@@ -0,0 +1,23 @@
+<!-- Uploaded to: SVG Repo, www.svgrepo.com, Generator: SVG Repo Mixer Tools -->
```
Comment (piyushmalviya13) [inline src/components/delete-button.svelte]: Create a core-commons folder, and an icons sub folder inside it and move all icons inside that

File: src/lib/models/File.ts [hunk 9]
```diff
@@ -0,0 +1,95 @@
+import { deleteUser, getAuth } from "firebase/auth";
+import { auth, db } from "../firebase/firebase";
+import { user } from "../store/store";
+
+export class File {
```
Comment (piyushmalviya13) [inline src/lib/models/File.ts]: Logic to name it File?

File: src/lib/models/File.ts [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    if (!File.instance) {
+      File.instance = new File();
+    }
+    return File.instance;
+  }
+
+  public async createUser(email: string, password: string) {
+    try{
+      const cred = await auth().createUserWithEmailAndPassword(email, password)
+      db.collection("users").doc(cred?.user?.uid).set({
```
Comment (piyushmalviya13) [inline src/lib/models/File.ts]: No checks for error in response. 
Straight away accessing db for user data?

File: src/lib/models/File.ts [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    } catch(err:any){
+      alert(err.message);
+      return false;
+    }
+  }
+
+  public async signInUser(email: string, password: string) {
+    try{
+      const cred:any = await auth().signInWithEmailAndPassword(email, password);
+      localStorage.setItem("userId", cred.user?.uid);
```
Comment (piyushmalviya13) [inline src/lib/models/File.ts]: No checks for error in response.
Straight away setting local storage?

File: src/lib/models/File.ts [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+      });
+      return true;
+    } catch(err:any){
+      alert(err.message);
+      return false;
+    }
+  }
+
+  public getUserTasks(currentUserId: string) {
+    const data = db.collection("users").doc(currentUserId);
```
Comment (piyushmalviya13) [inline src/lib/models/File.ts]: users string should be a constant.
It is used at multiple places.

File: src/components/TaskTable.svelte (hunk 2 above)
Comment (piyushmalviya13) [reply src/components/TaskTable.svelte]: Remove the otherTasks variable. It is not required

File: src/lib/store/store.ts [hunk 13]
```diff
@@ -0,0 +1,16 @@
+import { writable, readable, type Writable } from "svelte/store";
+
+export let toDoList: any = writable([]);
+export let doneList: any = writable([]);
+
+
+
```
Comment (piyushmalviya13) [inline src/lib/store/store.ts]: formatting

File: src/routes/+page.svelte [hunk 14]
```diff
@@ -0,0 +1,100 @@
+<script lang="ts">
```
Comment (piyushmalviya13) [inline src/routes/+page.svelte]: Put register page under the correct path

File: src/routes/home/+page.svelte [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+	import TaskTable from '../../components/TaskTable.svelte';
+	import {toDoList, doneList, user} from '../../lib/store/store';
+	import { database } from '../../utils/utils';
+	import { v4 as uuidv4 } from "uuid";
+	import { goto } from '$app/navigation';
+	
+	let task = '',
+	error = "",
+	list:any = [], 
+	otherList:any = [],
```
Comment (piyushmalviya13) [inline src/routes/home/+page.svelte]: naming?

File: src/routes/home/+page.svelte [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+	loading = true,
+	currentUser:any,
+	currentUserId:any;
+
+	$: onMount(async ()=>{
```
Comment (piyushmalviya13) [inline src/routes/home/+page.svelte]: Remove $:

File: src/routes/home/+page.svelte [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+				list = [];
+				otherList = [];
+				toDoList.set([]);
+                doneList.set([]);
+				database.updateUserTasks(currentUser);
+            }
+        } else{
+			error = '*Note : No task in to-do list! Kindly add a task first.';
+        }
+    }
```
Comment (piyushmalviya13) [inline src/routes/home/+page.svelte]: correct code formatting

File: src/routes/home/+page.svelte [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+			<button
+				class="md:mx-4 btn btn-secondary max-md:btn-sm"
+				on:click={() => {
+					addTask();
+				}}>Add task</button
+			>
+			<button
+				class="md:mx-4 btn btn-error max-md:btn-sm"
+				on:click={() => removeAllTasks()}>Remove All</button
+			>
```
Comment (piyushmalviya13) [inline src/routes/home/+page.svelte]: Create components for each component

File: src/utils/utils.ts [hunk 19]
```diff
@@ -0,0 +1,27 @@
+import { File } from "$lib/models/File";
+export const database = File.getInstance();
+
+export const emailRegex = /^([a-zA-Z0-9._%-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,})$/;
```
Comment (piyushmalviya13) [inline src/utils/utils.ts]: Move all constants into a separate const file

**Review by piyushmalviya13 — CHANGES_REQUESTED** [review]
Looks into the comments

File: src/components/Navbar.svelte (hunk 1 above)
Comment (ashwath462) [reply src/components/Navbar.svelte]: Done

File: src/components/TaskTable.svelte (hunk 4 above)
Comment (ashwath462) [reply src/components/TaskTable.svelte]: Naming correction done.

File: src/components/TaskTable.svelte (hunk 3 above)
Comment (ashwath462) [reply src/components/TaskTable.svelte]: Done

File: src/components/TaskTable.svelte (hunk 5 above)
Comment (ashwath462) [reply src/components/TaskTable.svelte]: Resolved.

File: src/routes/home/+page.svelte (hunk 15 above)
Comment (ashwath462) [reply src/routes/home/+page.svelte]: resolved

File: src/routes/home/+page.svelte (hunk 16 above)
Comment (ashwath462) [reply src/routes/home/+page.svelte]: Done.

File: src/routes/home/+page.svelte (hunk 17 above)
Comment (ashwath462) [reply src/routes/home/+page.svelte]: Done.

File: src/utils/utils.ts (hunk 19 above)
Comment (ashwath462) [reply src/utils/utils.ts]: Done.

File: src/components/TaskTable.svelte (hunk 2 above)
Comment (ashwath462) [reply src/components/TaskTable.svelte]: Done.

File: src/lib/models/File.ts (hunk 9 above)
Comment (ashwath462) [reply src/lib/models/File.ts]: Resolved

File: src/components/TaskTable.svelte (hunk 6 above)
Comment (ashwath462) [reply src/components/TaskTable.svelte]: Done.

File: src/components/TaskTable.svelte (hunk 2 above)
Comment (ashwath462) [reply src/components/TaskTable.svelte]: Done.

File: src/components/TaskTable.svelte (hunk 7 above)
Comment (ashwath462) [reply src/components/TaskTable.svelte]: Done.

File: src/lib/store/store.ts (hunk 13 above)
Comment (ashwath462) [reply src/lib/store/store.ts]: Done.

File: src/routes/+page.svelte (hunk 14 above)
Comment (ashwath462) [reply src/routes/+page.svelte]: Done.

File: src/components/delete-button.svelte (hunk 8 above)
Comment (ashwath462) [reply src/components/delete-button.svelte]: Done.

File: src/lib/models/File.ts (hunk 12 above)
Comment (ashwath462) [reply src/lib/models/File.ts]: Done.

File: src/routes/home/+page.svelte (hunk 18 above)
Comment (ashwath462) [reply src/routes/home/+page.svelte]: Done.

File: src/lib/models/File.ts (hunk 11 above)
Comment (ashwath462) [reply src/lib/models/File.ts]: Catches the error in catch block.

File: src/lib/models/File.ts (hunk 10 above)
Comment (ashwath462) [reply src/lib/models/File.ts]: throws the same error in catch block.

**Pull request closed without merging**